relay-server/src/actors/processor.rs

```diff
@@ -936,68 +936,295 @@ impl EnvelopeProcessorService {
 
-        if state.has_event() {
-            event::scrub(state)?;
-            event::serialize(state)?;
-            if_processing!({
-                span::extract_from_event(state);
-            });
-        }
+                event::finalize(state, &self.inner.config)?;
+                self.light_normalize_event(state)?;
+                dynamic_sampling::normalize(state);
+                event::filter(state)?;
+                dynamic_sampling::run(state, &self.inner.config);
+
+                // We avoid extracting metrics if we are not sampling the event while in non-processing
+                // relays, in order to synchronize rate limits on indexed and processed transactions.
+                if self.inner.config.processing_enabled() || state.sampling_result.should_drop() {
+                    self.extract_metrics(state)?;
+                }
+
+                dynamic_sampling::sample_envelope(state)?;
+
+                if_processing!({
+                    event::store(state, &self.inner.config, self.inner.geoip_lookup.as_ref())?;
+                    self.enforce_quotas(state)?;
+                    profile::process(state, &self.inner.config);
+                });
+
+                if state.has_event() {
+                    event::scrub(state)?;
+                    event::serialize(state)?;
+                    if_processing!({
+                        span::extract_from_event(state);
+                    });
+                }
+
+                attachment::scrub(state);
+            }
+            // Standalone attachments.
+            ItemType::Attachment | ItemType::FormData => {
+                // Only some attachments types have create event set to true.
+                if state.creates_event() {
+                    event::extract(state, &self.inner.config)?;
+
+                    if_processing!({
+                        attachment::create_placeholders(state);
+                    });
+
+                    event::finalize(state, &self.inner.config)?;
+                    self.light_normalize_event(state)?;
+                    event::filter(state)?;
+
+                    if_processing!({
+                        event::store(state, &self.inner.config, self.inner.geoip_lookup.as_ref())?;
+                    });
+                }
+                if_processing!({
+                    self.enforce_quotas(state)?;
+                });
+
+                if state.has_event() {
+                    event::scrub(state)?;
+                    event::serialize(state)?;
+                }
+
+                attachment::scrub(state);
+            }
+            ItemType::Session | ItemType::Sessions => {
+                session::process(state, &self.inner.config);
+                if_processing!({
+                    self.enforce_quotas(state)?;
+                });
+
+                if state.has_event() {
+                    event::scrub(state)?;
+                    event::serialize(state)?;
+                }
```

Comment:
```suggestion
```

Like @Dav1dde said.

Reply:
@jjbayer  the sessions will never have an event ?

Reply:
Every item that does not `require_event` is split into an event-less envelope in `queue_envelope` (see linked code), so yes I think that is safe to assume.

https://github.com/getsentry/relay/blob/e655ae18be1c4aa20742bc687a7e768b92ad2b9e/relay-server/src/endpoints/common.rs#L306-L322